### DISCUSSION TITLE
configuration to generate QR codes

### DISCUSSION THREAD

**Discussion by geeseven** [qutebrowser/qutebrowser#5927] [Show and tell] [answered]
From time to time, I need to create a QR code to quickly get a URL to a mobile device. I came up with two different configuration approaches.  

```python
config.bind('Q', ':open --tab https://duckduckgo.com/?q=!qr+{url}')
```

This takes the current URL, uses DuckDuckGo's QR code generation tool, and displays it in a new tab.

```python
config.bind('Q', 'hint links spawn /path/to/qutebrowser-qr-code.sh {hint-url}')
```

```bash
$ cat /path/to/qutebrowser-qr-code.sh
#!/usr/bin/env sh
filename=/tmp/qr.txt
qrencode --type=UTF8 --output="$filename" "$1"
echo "Press Q/q to quit" >> "$filename"
termite --exec="less '$filename'"
rm "$filename"
```

This approach is local and requires the `qrencode` command.  It works by 'Q' bringing up hints, then the selected URL gets a text file created with an ASCII QR code.  Next, the QR code will get displayed via `less` in a terminal, and finally deletes the file once you exit `less`.  You'll need to change the next to last line if you don't use the termite terminal.  I wish `qrencode` had a way to print ASCII to stdout instead of only being able to write a file.  

I hope someone might find these usefully or as a starting point to make something better for your needs.

**Comment by toofar** [comment]
> I wish qrencode had a way to print ASCII to stdout instead of only being able to write a file.

`qrencode --type=UTF8 -o -  "$1"` seems to do it? Well, it isn't ascii.

**Reply by The-Compiler** [reply]
With a 256color terminal, `ANSI256UTF8` will give you a white (rather than your terminal fg color) background. You *can* use `-t ASCII` as well, but I doubt there's a reader which will scan it :smile: 

Also, I just discovered I could actually use [kitty's](https://sw.kovidgoyal.net/kitty/index.html) image support as well: `qrencode -o- -t PNG -s 10 "$1" | kitty +kitten icat --align left`

**Comment by The-Compiler** [comment, marked as answer]
Thanks for the inspiration! :+1: I [added](https://github.com/qutebrowser/qutebrowser/commit/6c42f977814a3bfe937235b269bb80a5ba81a7a5) a `qr` userscript to the repo which generates a PNG file with `qrencode` and opens it in qutebrowser directly.

**Comment by geeseven** [comment]
> qrencode --type=UTF8 -o - "$1" seems to do it?

I always forget about that. Thanks for the reminder.


> I added a qr userscript to the repo which generates a PNG file with qrencode and opens it in qutebrowser directly.

Nice.  My script was designed for more general purpose in mind.  For a userscript, the self contained PNG method makes sense.  

Any reason why `/usr/bin/env bash` was not used?

**Reply by The-Compiler** [reply]
> Any reason why `/usr/bin/env bash` was not used?

Not really, mostly me being used to writing `#!/bin/bash`. I fixed it and added an automated check in e07611fc31f8d10c41f6b135d2b9b4713d4feee6.
